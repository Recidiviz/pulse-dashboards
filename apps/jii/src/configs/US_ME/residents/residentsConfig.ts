// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { ResidentsConfig } from "../../types";
import { config } from "../usMeSCCP/config";
import releaseDateInfo from "./releaseDateInfo.md?raw";

export const usMeResidentsConfig: ResidentsConfig = {
  headerProfileFields: [
    {
      label: "Current release date",
      value:
        "{{#if resident.releaseDate}}{{formatFullDate resident.releaseDate}}{{else}}Not available{{/if}}",
      moreInfo: releaseDateInfo,
    },
  ],
  incarcerationOpportunities: {
    usMeSCCP: config,
  },
  home: {
    progress: {
      title: "Your Progress",
    },
    eligibility: {
      title: "Your Eligibility",
    },
    footer: {
      about: {
        title: "About this app",
        body: `This app is made by Recidiviz, a technology nonprofit.  
        It is free to access. The information updates daily.`,
      },
      contact: {
        title: "Contact us",
        body: `1322 Webster St, Suite 402  
        Oakland, CA 94612`,
      },
    },
  },
};
