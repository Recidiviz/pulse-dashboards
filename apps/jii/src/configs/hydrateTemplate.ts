// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================
import { toTitleCase } from "@artsy/to-title-case";
import { isFuture } from "date-fns";
import Handlebars from "handlebars";

import { dateStringSchema } from "~datatypes";

import { formatFullDate } from "../utils/date";

Handlebars.registerHelper("titleCase", (s: string) => toTitleCase(s));
Handlebars.registerHelper("lowerCase", (s: string) => s.toLowerCase());
Handlebars.registerHelper("formatFullDate", (d: Date | string) => {
  const dateToFormat = d instanceof Date ? d : dateStringSchema.parse(d);
  return formatFullDate(dateToFormat);
});
Handlebars.registerHelper("isFutureDate", (d: Date) => isFuture(d));
Handlebars.registerHelper("and", (a: unknown, b: unknown) => a && b);
Handlebars.registerHelper("equals", (a: unknown, b: unknown) => a === b);

/**
 * Hydrates the Handlebars template in `template` with the values in `context`.
 */
export function hydrateTemplate(
  template: string,
  context: Record<string, unknown>,
) {
  return Handlebars.compile(template)(context);
}

/**
 * Handlebars templates may have extra whitespace for readability;
 * this condenses all that to single spaces and a single line
 * with no leading or trailing spaces
 */
export function cleanupInlineTemplate(s: string) {
  return s.replaceAll(/\s+/g, " ").trim();
}
