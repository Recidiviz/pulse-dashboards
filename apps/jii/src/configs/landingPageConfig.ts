// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { LandingPageConfig } from "./types";

export const landingPageConfig: LandingPageConfig = {
  copy: {
    intro: `# Learn about opportunities you may be eligible for.`,
    selectorLabel: "Find opportunities in the state where you’re incarcerated",
    selectorPlaceholder: "Select a state …",
  },
  states: [{ stateCode: "US_ME", displayName: "Maine", urlSlug: "maine" }],
};
