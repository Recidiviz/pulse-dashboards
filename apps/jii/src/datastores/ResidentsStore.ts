// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import keyBy from "lodash/keyBy";
import { makeAutoObservable, set } from "mobx";

import { ResidentRecord } from "~datatypes";
import { FilterParams } from "~firestore-api";
import { FlowMethod } from "~hydration-utils";

import { DataAPI } from "../apis/data/interface";
import { residentEligibilityReportConstructors } from "../configs/residentEligibilityReportConstructors";
import {
  IncarcerationOpportunityId,
  OpportunityRecord,
  ResidentsConfig,
} from "../configs/types";
import { EligibilityReport } from "../models/EligibilityReport/interface";
import type { RootStore } from "./RootStore";

// because we use a mapped type to ensure key-value agreement, this can't be an ES6 Map
type OpportunityRecordMapping = {
  [O in IncarcerationOpportunityId]?: OpportunityRecord<O>;
};

export class ResidentsStore {
  /**
   * Holds all resident records that have been fetched
   */
  residentsByExternalId: Map<string, ResidentRecord["output"]> = new Map();

  /**
   * Holds all opportunity eligibility records that have been fetched
   */
  residentOpportunityRecordsByExternalId: Map<
    string,
    OpportunityRecordMapping
  > = new Map();

  residentEligibilityReportsByExternalId: Map<
    string,
    Map<IncarcerationOpportunityId, EligibilityReport>
  > = new Map();

  constructor(
    private readonly rootStore: RootStore,
    public readonly config: ResidentsConfig,
  ) {
    makeAutoObservable(this);
  }

  private get apiClient() {
    return this.rootStore.apiClient;
  }

  get userStore() {
    return this.rootStore.userStore;
  }

  get opportunityIdsByUrlSlug(): Map<string, IncarcerationOpportunityId> {
    return new Map(
      Object.entries(this.config.incarcerationOpportunities).map(
        ([id, config]) => {
          return [
            config.urlSlug,
            id as keyof typeof this.config.incarcerationOpportunities,
          ];
        },
      ),
    );
  }

  /**
   * Get a convenient non-nullable ID value; accepts any string,
   * so you should be confident you have a valid slug or be prepared to handle an error.
   */
  opportunitySlugToIdOrThrow(slug: string) {
    const id = this.opportunityIdsByUrlSlug.get(slug);
    if (!id) {
      throw new Error(`No opportunity ID matches url segment ${slug}`);
    }
    return id;
  }

  areAllResidentsPopulated(): boolean {
    // if we started the session on a single resident's page, we might have had one
    // populated already. Seems a safe assumption that the total will always be > 1
    return this.residentsByExternalId.size > 1;
  }
  /**
   * Populates {@link residentsByExternalId} with the API response for all available residents.
   * Will not refetch if data is already populated, unless `forceRefresh` is true
   */
  *populateResidents(
    filters?: Array<FilterParams>,
    forceRefresh = false,
  ): FlowMethod<DataAPI["residents"], void> {
    if (!forceRefresh && this.areAllResidentsPopulated()) return;

    const residents = yield this.apiClient.residents(filters);
    set(this.residentsByExternalId, keyBy(residents, "personExternalId"));
  }

  isResidentPopulated(residentExternalId: string): boolean {
    return this.residentsByExternalId.has(residentExternalId);
  }

  /**
   * Populates {@link residentsByExternalId} with the API response for the resident
   * with personExternalId matching `residentExternalId`. Throws if the API request fails.
   * Will not refetch if data is already populated.
   */
  *populateResidentById(
    residentExternalId: string,
  ): FlowMethod<DataAPI["residentById"], void> {
    if (this.isResidentPopulated(residentExternalId)) return;

    const resident = yield this.apiClient.residentById(residentExternalId);
    this.residentsByExternalId.set(resident.personExternalId, resident);
  }

  get residentsByPseudoId(): Map<string, ResidentRecord["output"]> {
    return new Map(
      Object.entries(
        keyBy([...this.residentsByExternalId.values()], "pseudonymizedId"),
      ),
    );
  }

  isResidentWithPseudoIdPopulated(pseudoId: string): boolean {
    return this.residentsByPseudoId.has(pseudoId);
  }

  /**
   * Populates {@link residentsByExternalId} with the API response for the resident
   * with pseudonymizedId matching `residentPseudoId`. Throws if the API request fails.
   * Will not refetch if data is already populated.
   */
  *populateResidentByPseudoId(
    residentPseudoId: string,
  ): FlowMethod<DataAPI["residentById"], void> {
    if (this.isResidentWithPseudoIdPopulated(residentPseudoId)) return;

    const resident = yield this.apiClient.residentByPseudoId(residentPseudoId);
    this.residentsByExternalId.set(resident.personExternalId, resident);
  }

  isResidentEligibilityRecordPopulated(
    residentExternalId: string,
    opportunityId: IncarcerationOpportunityId,
  ): boolean {
    return (
      opportunityId in
      (this.residentOpportunityRecordsByExternalId.get(residentExternalId) ??
        {})
    );
  }

  /**
   * Populates {@link residentOpportunityRecordsByExternalId} with the API response for
   * the provided resident and opportunity IDs. Will populate with `undefined` if no record is found
   * (indicating the resident is ineligible). Will not refetch if data is already populated.
   */
  *populateOpportunityRecordByResidentId(
    residentExternalId: string,
    opportunityId: IncarcerationOpportunityId,
  ): FlowMethod<DataAPI["residentEligibility"], void> {
    if (
      this.isResidentEligibilityRecordPopulated(
        residentExternalId,
        opportunityId,
      )
    )
      return;

    const eligibilityRecord = yield this.apiClient.residentEligibility(
      residentExternalId,
      opportunityId,
    );
    const opportunityMap: OpportunityRecordMapping =
      this.residentOpportunityRecordsByExternalId.get(residentExternalId) ?? {};

    opportunityMap[opportunityId] = eligibilityRecord;

    if (!this.residentOpportunityRecordsByExternalId.has(residentExternalId)) {
      this.residentOpportunityRecordsByExternalId.set(
        residentExternalId,
        opportunityMap,
      );
    }
  }

  isResidentEligibilityReportPopulated(
    residentExternalId: string,
    opportunityId: IncarcerationOpportunityId,
  ): boolean {
    return !!this.residentEligibilityReportsByExternalId
      .get(residentExternalId)
      ?.has(opportunityId);
  }

  /**
   * Populates {@link residentEligibilityReportsByExternalId} with an {@link EligibilityReport} object
   * for the given resident and opportunity, if it does not already exist.
   */
  populateEligibilityReportFromData<OppId extends IncarcerationOpportunityId>(
    opportunityId: OppId,
    resident: ResidentRecord["output"],
    opportunity: OpportunityRecord<OppId> | undefined,
  ) {
    const residentExternalId = resident.personExternalId;

    if (
      this.isResidentEligibilityReportPopulated(
        residentExternalId,
        opportunityId,
      )
    ) {
      return;
    }

    const config = this.config.incarcerationOpportunities[opportunityId];

    if (!config) {
      throw new Error(`Opportunity ${opportunityId} is not configured`);
    }

    const ReportConstructor =
      residentEligibilityReportConstructors[opportunityId];

    const mapping =
      this.residentEligibilityReportsByExternalId.get(residentExternalId) ??
      new Map();

    mapping.set(
      opportunityId,
      new ReportConstructor(resident, config, opportunity),
    );
    if (!this.residentEligibilityReportsByExternalId.has(residentExternalId)) {
      this.residentEligibilityReportsByExternalId.set(
        residentExternalId,
        mapping,
      );
    }
  }

  get stateCode() {
    return this.userStore.stateCode;
  }
}
