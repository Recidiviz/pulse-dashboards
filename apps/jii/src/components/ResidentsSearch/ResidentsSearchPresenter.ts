// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { flowResult, makeAutoObservable } from "mobx";

import { isDemoMode, isOfflineMode } from "~client-env-utils";
import { ResidentRecord } from "~datatypes";
import { FilterParams } from "~firestore-api";
import {
  Hydratable,
  HydratesFromSource,
  HydrationState,
} from "~hydration-utils";

import { ResidentsStore } from "../../datastores/ResidentsStore";
import { UiStore } from "../../datastores/UiStore";

const ALL_FACILITIES_VALUE = "__ALL__";

type SelectOption = { label: string; value: string };

export class ResidentsSearchPresenter implements Hydratable {
  private hydrationSource: HydratesFromSource;

  constructor(
    private residentsStore: ResidentsStore,
    private uiStore: UiStore,
  ) {
    makeAutoObservable(this, undefined, { autoBind: true });

    this.hydrationSource = new HydratesFromSource({
      populate: async () => {
        await flowResult(
          this.residentsStore.populateResidents(this.residentsFilterParams),
        );
      },
      expectPopulated: [this.expectResidentsPopulated],
    });
  }

  private expectResidentsPopulated() {
    if (!this.residentsStore.areAllResidentsPopulated()) {
      throw new Error("Residents data is not populated");
    }
  }

  get hydrationState(): HydrationState {
    return this.hydrationSource.hydrationState;
  }

  async hydrate(): Promise<void> {
    return this.hydrationSource.hydrate();
  }

  private get residents(): Array<ResidentRecord["output"]> {
    const { selectedFacilityFilterOptionValue } = this.uiStore;

    return Array.from(
      this.residentsStore.residentsByExternalId.values(),
    ).filter(
      (r) =>
        !selectedFacilityFilterOptionValue ||
        selectedFacilityFilterOptionValue === ALL_FACILITIES_VALUE ||
        r.facilityId === selectedFacilityFilterOptionValue,
    );
  }

  get selectOptions() {
    return this.residents.map((r) => ({
      value: r,
      label: `${r.personName.givenNames} ${r.personName.surname} (${r.personExternalId})`,
    }));
  }

  get facilityFilterOptions(): [SelectOption, ...Array<SelectOption>] {
    const options: [SelectOption, ...Array<SelectOption>] = [
      { label: "All", value: ALL_FACILITIES_VALUE },
    ];

    // for convenience this is just hardcoded for the pilot
    if (!isOfflineMode() && !isDemoMode()) {
      options.push({
        label: "Mountain View Correctional Facility",
        value: "MOUNTAIN VIEW CORRECTIONAL FACILITY",
      });
    }

    return options;
  }

  /**
   * The facility select component is uncontrolled, but this can be used preserve state when navigating
   * away from the page (passing a default only affects which option is selected when the component mounts)
   */
  get facilityFilterDefaultOption() {
    const { selectedFacilityFilterOptionValue } = this.uiStore;
    if (selectedFacilityFilterOptionValue) {
      return this.facilityFilterOptions.find(
        (o) => o.value === selectedFacilityFilterOptionValue,
      );
    }

    return this.facilityFilterOptions.at(-1);
  }

  private get residentsFilterParams(): Array<FilterParams> | undefined {
    const { facilityFilterDefaultOption } = this;
    if (!facilityFilterDefaultOption) return;

    if (facilityFilterDefaultOption.value === ALL_FACILITIES_VALUE) return;

    return [["facilityId", "==", facilityFilterDefaultOption.value]];
  }

  setFacilityFilter(value: string) {
    if (value !== this.uiStore.selectedFacilityFilterOptionValue) {
      this.uiStore.selectedFacilityFilterOptionValue = value;
      this.residentsStore.populateResidents(this.residentsFilterParams, true);
    }
  }
}
