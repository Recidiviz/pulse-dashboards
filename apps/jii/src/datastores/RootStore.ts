// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================
import { captureException } from "@sentry/react";
import { makeObservable, onReactionError } from "mobx";

import { isOfflineMode, isTestEnv } from "~client-env-utils";
import { FlowMethod } from "~hydration-utils";

import { ApiClient } from "../apis/data/ApiClient";
import { DataAPI } from "../apis/data/interface";
import { OfflineAPIClient } from "../apis/data/OfflineAPIClient";
import { StateCode } from "../configs/types";
import { ResidentsStore } from "./ResidentsStore";
import { UiStore } from "./UiStore";
import { UserStore } from "./UserStore";

// global error handling for Mobx reactions
onReactionError((error) => {
  captureException(error);
});

export class RootStore {
  /**
   * Will be an offline client in offline and test modes,
   * and an external API client otherwise.
   */
  apiClient: DataAPI;

  residentsStore?: ResidentsStore;

  userStore: UserStore;

  uiStore: UiStore;

  // for convenience, this is a constant while we only have one state onboarded
  readonly stateCode: StateCode = "US_ME";

  constructor() {
    makeObservable(this, {
      populateResidentsStore: true,
      residentsStore: true,
    });

    this.userStore = new UserStore(this);

    this.uiStore = new UiStore();

    this.apiClient = this.createApiClient();
  }

  /**
   * Constructs and returns an appropriate API client based on the environment
   * (e.g. one that returns fixture data in offline mode). Not intended to be
   * MobX-reactive, since it creates a brand-new instance and discards any state
   * the previous instance might have held.
   */
  private createApiClient(): DataAPI {
    if (isOfflineMode() || isTestEnv()) {
      return new OfflineAPIClient(this);
    } else {
      // eslint-disable-next-line @typescript-eslint/no-this-alias
      const store = this;
      const externals = {
        // these are getters rather than static properties so that the API client
        // can can observe changes to these values rather than getting a one-time snapshot
        get stateCode() {
          return store.stateCode;
        },
        get authClient() {
          return store.userStore.authClient;
        },
        get config() {
          return store.residentsStore?.config;
        },
      };

      return new ApiClient(externals);
    }
  }

  *populateResidentsStore(): FlowMethod<DataAPI["residentsConfig"], void> {
    const config = yield this.apiClient.residentsConfig();

    this.residentsStore = new ResidentsStore(this, config);
  }
}
