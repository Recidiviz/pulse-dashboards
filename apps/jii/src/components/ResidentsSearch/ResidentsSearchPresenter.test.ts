// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { waitFor } from "@testing-library/react";
import { keyBy } from "lodash";
import { configure, set } from "mobx";

import { isDemoMode } from "~client-env-utils";
import { usMeResidents } from "~datatypes";

import { residentsConfigByState } from "../../configs/residentsConfig";
import { ResidentsStore } from "../../datastores/ResidentsStore";
import { RootStore } from "../../datastores/RootStore";
import { ResidentsSearchPresenter } from "./ResidentsSearchPresenter";

vi.mock("~client-env-utils", () => ({
  isTestEnv: () => true,
  isOfflineMode: vi.fn(),
  isDemoMode: vi.fn(),
}));

let residentsStore: ResidentsStore;
let presenter: ResidentsSearchPresenter;
let rootStore: RootStore;

beforeEach(() => {
  configure({ safeDescriptors: false });
  rootStore = new RootStore();
  vi.spyOn(
    rootStore.userStore.authClient,
    "appMetadata",
    "get",
  ).mockReturnValue({
    stateCode: "US_ME",
  });

  residentsStore = new ResidentsStore(rootStore, residentsConfigByState.US_ME);
  presenter = new ResidentsSearchPresenter(residentsStore, rootStore.uiStore);
});

afterEach(() => {
  configure({ safeDescriptors: true });
});

describe("hydration", () => {
  test("needs hydration", () => {
    expect(presenter.hydrationState.status).toBe("needs hydration");
    expect(presenter.selectOptions).toEqual([]);
  });

  test("already hydrated", () => {
    set(
      residentsStore.residentsByExternalId,
      keyBy(usMeResidents, "personExternalId"),
    );

    expect(presenter.hydrationState.status).toBe("hydrated");
  });

  test("hydrate", async () => {
    vi.spyOn(residentsStore, "populateResidents");
    expect(presenter.hydrationState.status).toBe("needs hydration");

    presenter.hydrate();

    expect(presenter.hydrationState.status).toBe("loading");

    await waitFor(() =>
      expect(presenter.hydrationState.status).toBe("hydrated"),
    );

    expect(presenter.selectOptions).toMatchSnapshot();
    expect(residentsStore.populateResidents).toHaveBeenLastCalledWith([
      ["facilityId", "==", "MOUNTAIN VIEW CORRECTIONAL FACILITY"],
    ]);
  });
});

describe("facility filter", () => {
  test("default value", () => {
    expect(presenter.facilityFilterDefaultOption).toMatchInlineSnapshot(`
      {
        "label": "Mountain View Correctional Facility",
        "value": "MOUNTAIN VIEW CORRECTIONAL FACILITY",
      }
    `);
  });

  test("set value", async () => {
    vi.spyOn(residentsStore, "populateResidents");

    presenter.setFacilityFilter("__ALL__");
    expect(presenter.facilityFilterDefaultOption).toMatchInlineSnapshot(`
      {
        "label": "All",
        "value": "__ALL__",
      }
    `);

    expect(residentsStore.populateResidents).toHaveBeenLastCalledWith(
      undefined,
      true,
    );

    presenter.setFacilityFilter("MOUNTAIN VIEW CORRECTIONAL FACILITY");

    expect(presenter.facilityFilterDefaultOption).toMatchInlineSnapshot(`
      {
        "label": "Mountain View Correctional Facility",
        "value": "MOUNTAIN VIEW CORRECTIONAL FACILITY",
      }
    `);

    expect(residentsStore.populateResidents).toHaveBeenLastCalledWith(
      [["facilityId", "==", "MOUNTAIN VIEW CORRECTIONAL FACILITY"]],
      true,
    );
  });

  test("demo mode", () => {
    vi.mocked(isDemoMode).mockReturnValue(true);

    expect(presenter.facilityFilterOptions).toEqual([
      {
        label: "All",
        value: "__ALL__",
      },
    ]);

    expect(presenter.facilityFilterDefaultOption).toEqual({
      label: "All",
      value: "__ALL__",
    });
  });
});
