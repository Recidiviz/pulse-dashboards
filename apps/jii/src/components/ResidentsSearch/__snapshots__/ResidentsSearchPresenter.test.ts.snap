// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`hydration > hydrate 1`] = `
[
  {
    "label": "First Resident (RES001)",
    "value": {
      "admissionDate": 2013-11-16T00:00:00.000Z,
      "custodyLevel": "MINIMUM",
      "displayId": "dRES001",
      "facilityId": "FACILITY NAME",
      "gender": "MALE",
      "officerId": "OFFICER5",
      "personExternalId": "RES001",
      "personName": {
        "givenNames": "First",
        "surname": "Resident",
      },
      "personType": "RESIDENT",
      "portionServedNeeded": "2/3",
      "pseudonymizedId": "anonres001",
      "releaseDate": 2024-11-16T00:00:00.000Z,
      "stateCode": "US_ME",
      "unitId": "UNIT A",
      "usMePortionNeededEligibleDate": 2021-11-16T00:00:00.000Z,
    },
  },
  {
    "label": "Second Resident (RES002)",
    "value": {
      "admissionDate": 2018-02-17T00:00:00.000Z,
      "custodyLevel": "MINIMUM",
      "displayId": "dRES002",
      "facilityId": "FACILITY NAME",
      "gender": "FEMALE",
      "officerId": "OFFICER5",
      "personExternalId": "RES002",
      "personName": {
        "givenNames": "Second",
        "surname": "Resident",
      },
      "personType": "RESIDENT",
      "portionServedNeeded": "2/3",
      "pseudonymizedId": "anonres002",
      "releaseDate": 2024-02-16T00:00:00.000Z,
      "stateCode": "US_ME",
      "unitId": "UNIT B",
      "usMePortionNeededEligibleDate": 2022-02-16T00:00:00.000Z,
    },
  },
  {
    "label": "Third Resident (RES003)",
    "value": {
      "admissionDate": 2019-06-18T00:00:00.000Z,
      "custodyLevel": "MINIMUM",
      "displayId": "dRES003",
      "facilityId": "FACILITY NAME",
      "gender": "TRANS_FEMALE",
      "officerId": "OFFICER5",
      "personExternalId": "RES003",
      "personName": {
        "givenNames": "Third",
        "surname": "Resident",
      },
      "personType": "RESIDENT",
      "portionServedNeeded": "1/2",
      "pseudonymizedId": "anonres003",
      "releaseDate": 2023-12-16T00:00:00.000Z,
      "stateCode": "US_ME",
      "unitId": "UNIT B",
      "usMePortionNeededEligibleDate": 2021-09-16T00:00:00.000Z,
    },
  },
  {
    "label": "Fourth Resident (RES004)",
    "value": {
      "admissionDate": 2019-05-18T00:00:00.000Z,
      "custodyLevel": "COMMUNITY",
      "displayId": "dRES004",
      "facilityId": "FACILITY NAME",
      "gender": "INTERNAL_UNKNOWN",
      "officerId": "OFFICER5",
      "personExternalId": "RES004",
      "personName": {
        "givenNames": "Fourth",
        "surname": "Resident",
      },
      "personType": "RESIDENT",
      "portionServedNeeded": "1/2",
      "pseudonymizedId": "anonres004",
      "releaseDate": 2024-01-16T00:00:00.000Z,
      "stateCode": "US_ME",
      "unitId": "UNIT A",
      "usMePortionNeededEligibleDate": 2021-09-16T00:00:00.000Z,
    },
  },
  {
    "label": "Fifth Resident (RES005)",
    "value": {
      "admissionDate": 2020-05-16T00:00:00.000Z,
      "custodyLevel": "MINIMUM",
      "displayId": "dRES005",
      "facilityId": "FACILITY NAME",
      "gender": "TRANS",
      "officerId": "OFFICER5",
      "personExternalId": "RES005",
      "personName": {
        "givenNames": "Fifth",
        "surname": "Resident",
      },
      "personType": "RESIDENT",
      "portionServedNeeded": "1/2",
      "pseudonymizedId": "anonres005",
      "releaseDate": 2024-05-16T00:00:00.000Z,
      "stateCode": "US_ME",
      "unitId": "UNIT B",
      "usMePortionNeededEligibleDate": 2022-05-16T00:00:00.000Z,
    },
  },
  {
    "label": "Sixth Resident (RES006)",
    "value": {
      "admissionDate": 2015-02-16T00:00:00.000Z,
      "custodyLevel": "MINIMUM",
      "displayId": "dRES006",
      "facilityId": "FACILITY NAME",
      "gender": "TRANS",
      "officerId": "OFFICER5",
      "personExternalId": "RES006",
      "personName": {
        "givenNames": "Sixth",
        "surname": "Resident",
      },
      "personType": "RESIDENT",
      "portionServedNeeded": "2/3",
      "pseudonymizedId": "anonres006",
      "releaseDate": 2024-02-16T00:00:00.000Z,
      "stateCode": "US_ME",
      "unitId": "UNIT B",
      "usMePortionNeededEligibleDate": 2021-02-16T00:00:00.000Z,
    },
  },
  {
    "label": "Seventh Resident (RES007)",
    "value": {
      "admissionDate": 2019-06-16T00:00:00.000Z,
      "custodyLevel": "MINIMUM",
      "displayId": "dRES007",
      "facilityId": "FACILITY NAME",
      "gender": "FEMALE",
      "officerId": "OFFICER5",
      "personExternalId": "RES007",
      "personName": {
        "givenNames": "Seventh",
        "surname": "Resident",
      },
      "personType": "RESIDENT",
      "portionServedNeeded": "1/2",
      "pseudonymizedId": "anonres007",
      "releaseDate": 2023-12-16T00:00:00.000Z,
      "stateCode": "US_ME",
      "unitId": "UNIT B",
      "usMePortionNeededEligibleDate": 2021-09-16T00:00:00.000Z,
    },
  },
  {
    "label": "Eighth Resident (RES008)",
    "value": {
      "admissionDate": 2015-08-16T00:00:00.000Z,
      "custodyLevel": "MINIMUM",
      "displayId": "dRES008",
      "facilityId": "FACILITY NAME",
      "gender": "TRANS",
      "officerId": "OFFICER5",
      "personExternalId": "RES008",
      "personName": {
        "givenNames": "Eighth",
        "surname": "Resident",
      },
      "personType": "RESIDENT",
      "portionServedNeeded": "2/3",
      "pseudonymizedId": "anonres008",
      "releaseDate": 2024-08-16T00:00:00.000Z,
      "stateCode": "US_ME",
      "unitId": "UNIT A",
      "usMePortionNeededEligibleDate": 2021-08-16T00:00:00.000Z,
    },
  },
  {
    "label": "Ninth Resident (RES009)",
    "value": {
      "admissionDate": 2014-12-16T00:00:00.000Z,
      "custodyLevel": "MINIMUM",
      "displayId": "dRES009",
      "facilityId": "FACILITY NAME",
      "gender": "MALE",
      "officerId": "OFFICER5",
      "personExternalId": "RES009",
      "personName": {
        "givenNames": "Ninth",
        "surname": "Resident",
      },
      "personType": "RESIDENT",
      "portionServedNeeded": "1/2",
      "pseudonymizedId": "anonres009",
      "releaseDate": 2025-02-16T00:00:00.000Z,
      "stateCode": "US_ME",
      "unitId": "UNIT A",
      "usMePortionNeededEligibleDate": 2020-01-16T00:00:00.000Z,
    },
  },
  {
    "label": "NoRelease Resident (RES998)",
    "value": {
      "admissionDate": 2020-05-16T00:00:00.000Z,
      "custodyLevel": "MEDIUM",
      "displayId": "dRES998",
      "facilityId": "FACILITY NAME",
      "gender": "MALE",
      "officerId": "OFFICER5",
      "personExternalId": "RES998",
      "personName": {
        "givenNames": "NoRelease",
        "surname": "Resident",
      },
      "personType": "RESIDENT",
      "pseudonymizedId": "anonres998",
      "stateCode": "US_ME",
      "unitId": "UNIT B",
    },
  },
  {
    "label": "Ineligible Resident (RES999)",
    "value": {
      "admissionDate": 2020-05-16T00:00:00.000Z,
      "custodyLevel": "MEDIUM",
      "displayId": "dRES999",
      "facilityId": "FACILITY NAME",
      "gender": "MALE",
      "officerId": "OFFICER5",
      "personExternalId": "RES999",
      "personName": {
        "givenNames": "Ineligible",
        "surname": "Resident",
      },
      "personType": "RESIDENT",
      "portionServedNeeded": "1/2",
      "pseudonymizedId": "anonres999",
      "releaseDate": 2025-05-16T00:00:00.000Z,
      "stateCode": "US_ME",
      "unitId": "UNIT B",
      "usMePortionNeededEligibleDate": 2022-11-16T00:00:00.000Z,
    },
  },
]
`;
